sklearn/utils/set_output.py

```diff
@@ -0,0 +1,225 @@
+from functools import wraps
+
+from scipy.sparse import issparse
+
+from sklearn.utils import check_pandas_support
+from sklearn._config import get_config
+from sklearn.utils import Bunch
+
+__all__ = [
+    "get_output_config",
+    "make_named_container",
+    "SetOutputMixin",
+    "safe_set_output",
+]
+
+
+def make_named_container(
+    output,
+    *,
+    index=None,
+    columns=None,
+    dense_container="default",
+    constructor_kwargs=None,
+):
+    """Create a named container.
+
+    Parameters
+    ----------
+    output : ndarray, sparse matrix or pandas DataFrame
+        Container to name.
+
+    index : array-like, default=None
+        Index for data.
+
+    columns : callable or ndarray, default=None
+        The column names or a callable that returns the column names. This is
+        useful if the column names require some computation.
+
+    dense_container : {"default", "pandas"}, default="default"
+        Container used for dense data.
+
+    constructor_kwargs : dict, default=None
+        Keyword arguments passed to container constructor.
+
+    Returns
+    -------
+    output : DataFrame or ndarray
+        Container with column names or unchanged `output`.
+    """
+    if dense_container not in {"default", "pandas"}:
+        raise ValueError(
+            f"dense_container must be 'default' or 'pandas' got {dense_container}"
+        )
+
+    if dense_container == "default":
+        return output
+
+    constructor_kwargs = constructor_kwargs or {}
+
+    # dense_container == "pandas"
+    if issparse(output):
+        raise ValueError("Pandas output does not support sparse data")
+
+    if callable(columns):
+        columns = columns()
+
+    if hasattr(output, "columns"):
+        if columns is not None:
+            output.columns = columns
+        if index is not None:
+            output.index = index
+        return output
```

Comment:
> should it be the responsibility of the implementer of that transformer to ensure consistency with get_feature_names_out

I'll prefer `get_feature_names_out` to already be consistent when it gets to wrapping the container. In that case, we can have `SetOutputMixin` be a no-op if the return of `transform` is already a DataFrame.

> Or is it just to add some compat with third-party estimators that output dataframes by default in case they plan to inherit from the SetOutputMixin?

This `if-block` is for compat, but it's also to make sense of the API. If the `output` is already a dataframe and `columns` is passed in, then I thought a reasonable thing to do is to set the `columns` as well. Given the point, above I am okay with doing a no-op if the container is already a DataFrame.

Reply:
I updated the PR to make it a noop when the `original_data` is a pandas dataframe. This way `SetOutputMixin` does not do anything if `estimator.transform` already returns a DataFrame.

Reply:
Is this resolved?

Reply:
I think there is still a lingering issue with `FunctionTransformer`.

This PR has been updated so that `TranformerMixin` inherits the `SetOutputMixin`. This means `FunctionTransformer` will get the `set_output` API if `feature_names_out` is defined. To be concrete, here is the current behavior for `FunctionTransformer`:

1. If `feature_names_out` is defined, then `set_output` is available. The output columns will always be consistent with `get_feature_names_out` no matter if `func` returns a dataframe or ndarray.

2. If `feature_names_out` is not defined then `set_output` is **not** available. `func` is allow to return a dataframe or a ndarray. Not having `set_output` is inconvenient, specifically when `func` returns a dataframe. For example, if the `FunctionTransformer` without `set_output` is in complex pipeline, then `pipeline.set_output(transform="pandas")` would fail to configure `FunctionTransformer`.

A workaround for 2 is to have a custom `set_output` raise a warning when `feature_names_out` is not defined. The warning states that `func` must return a dataframe if `set_output(transform="pandas")`. Technically, we can "learn" the feature names out by running `func` in `fit`, but that introduces more computation compared to `main`.

Reply:
> A workaround for 2 is to have a custom set_output raise a warning when feature_names_out is not defined. The warning states that func must return a dataframe if set_output(transform="pandas").

Sounds reasonable. We can also raise an error at transform time if `set_ouput(transform="pandas")` was previously called and that `func` does not naturally return a pandas dataframe.